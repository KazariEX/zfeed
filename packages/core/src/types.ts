export interface Item {
    title: string;
    id?: string;
    link: string;
    date: Date;
    published?: Date;

    description?: string;
    category?: Category[];
    content?: string;

    author?: Author[];
    contributor?: Author[];
    copyright?: string;

    enclosure?: Enclosure;
    image?: string | Enclosure;
    audio?: string | Enclosure;
    video?: string | Enclosure;

    extensions?: Extension[];
}

export interface Category {
    name?: string;
    domain?: string;
    scheme?: string;
    term?: string;
}

export interface Enclosure {
    url: string;
    type?: string;
    length?: number;
    title?: string;
    duration?: number;
}

export interface Author {
    name?: string;
    email?: string;
    link?: string;
    avatar?: string;
}

export interface Extension {
  name: string;
  objects: any;
}
