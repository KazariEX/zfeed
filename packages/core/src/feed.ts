import type { Author, Extension, Item } from "./types";

export interface Feed {
    title: string;
    description?: string;
    id: string;
    link?: string;
    feed?: string;
    feedLinks?: {
        atom?: string;
        json?: string;
        rss?: string;
    };
    hub?: string;

    language?: string;
    generator?: string;
    stylesheet?: string;
    docs?: string;
    image?: string;
    favicon?: string;
    copyright?: string;
    updated?: Date;
    ttl?: number;

    category?: string;
    podcast?: boolean;

    author?: Author;
    contributors: Author[];
    categories: string[];
    items: Item[];
    extensions: Extension[];
}

export interface CreateFeedOptions extends Partial<Feed> {
    title: string;
    id: string;
}

export function createFeed(options: CreateFeedOptions) {
    const feed: Feed = {
        items: [],
        categories: [],
        contributors: [],
        extensions: [],
        ...options,
    };
    return feed;
}
