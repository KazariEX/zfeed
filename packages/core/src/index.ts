export * from "./feed";
export * from "./generators/atom1";
export * from "./generators/json1";
export * from "./generators/rss2";
export * from "./plugin";
export * from "./types";
