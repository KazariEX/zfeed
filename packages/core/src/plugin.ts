import type { Plugin } from "./types";

export function definePlugin<T>(plugin: Plugin<T>) {
    return plugin;
}
