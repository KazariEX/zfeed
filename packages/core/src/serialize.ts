export function serialize(name: string, val: unknown): string {
    if (isNullish(val)) {
        return "";
    }
    if (Array.isArray(val)) {
        return val.map((item) => serialize(name, item)).join("");
    }

    let stuff = "";
    let content = "";
    let raw = false;

    if (typeof val === "object") {
        for (const [key, value] of Object.entries(val)) {
            if (isNullish(value)) {
                continue;
            }
            if (key.startsWith("$")) {
                stuff += ` ${key.slice(1)}="${escape(String(value))}"`;
            }
            else if (key === "#text") {
                content = escape(String(value));
                raw = true;
            }
            else if (key === "#cdata") {
                content = `<![CDATA[${value}]]>`;
                raw = true;
            }
            else if (!raw) {
                content += serialize(key, value);
            }
        }
    }
    else {
        content = escape(String(val));
    }

    return name
        ? name.startsWith("?")
            ? `<?${name.slice(1)}${stuff}?>`
            : content
                ? `<${name}${stuff}>${content}</${name}>`
                : `<${name}${stuff} />`
        : content;
}

const entities: Record<string, string> = {
    "&": "&amp;",
    "<": "&lt;",
    ">": "&gt;",
    "'": "&#39;",
    "\"": "&quot;",
};

function escape(str: string) {
    return str.replaceAll(/[&<>'"]/g, (match) => entities[match]);
}

function isNullish(val: unknown) {
    return val === null || val === void 0;
}
