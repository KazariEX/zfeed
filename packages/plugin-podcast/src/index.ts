import { definePlugin } from "zfeed";
import * as packageJson from "../package.json";

/**
 * @see https://podcast-standard.org/podcast_standard
 */
export default definePlugin(() => {
    return {
        name: packageJson.name,
        resolveRss2(feed, xml) {
            xml.rss["$xmlns:googleplay"] = "http://www.google.com/schemas/play-podcasts/1.0";
            xml.rss["$xmlns:itunes"] = "http://www.itunes.com/dtds/podcast-1.0.dtd";

            const author = Array.isArray(feed.author) ? feed.author[0] : feed.author;

            if (author?.name !== void 0) {
                xml.rss.channel["googleplay:author"] = author.name;
                xml.rss.channel["itunes:author"] = author.name;
            }

            if (author?.email !== void 0) {
                xml.rss.channel["googleplay:owner"] = author.email;
                xml.rss.channel["itunes:owner"] = {
                    "itunes:email": author.email,
                };
            }

            if (feed.categories !== void 0) {
                xml.rss.channel["googleplay:category"] = feed.categories.map(({ term }) => ({ $text: term }));
                xml.rss.channel["itunes:category"] = feed.categories.map(({ term }) => ({ $text: term }));
            }

            if (feed.image !== void 0) {
                xml.rss.channel["googleplay:image"] = { $href: feed.image };
                xml.rss.channel["itunes:image"] = { $href: feed.image };
            }
        },
    };
});
