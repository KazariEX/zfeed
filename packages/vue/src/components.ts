import { h, type SetupContext } from "vue";
import type { XSLApplyImportElementAttributes, XSLApplyTemplatesElementAttributes, XSLAttributeElementAttributes, XSLAttributeSetElementAttributes, XSLCallTemplateElementAttributes, XSLChooseElementAttributes, XSLCommentElementAttributes, XSLCopyElementAttributes, XSLCopyOfElementAttributes, XSLDecimalFormatElementAttributes, XSLElementElementAttributes, XSLFallbackElementAttributes, XSLForEachElementAttributes, XSLIfElementAttributes, XSLImportElementAttributes, XSLIncludeElementAttributes, XSLKeyElementAttributes, XSLMessageElementAttributes, XSLNamespaceAliasElementAttributes, XSLNumberElementAttributes, XSLOtherwiseElementAttributes, XSLOutputElementAttributes, XSLParamElementAttributes, XSLPreserveSpaceElementAttributes, XSLProcessingInstructionElementAttributes, XSLSortElementAttributes, XSLStripSpaceElementAttributes, XSLStylesheetElementAttributes, XSLTemplateElementAttributes, XSLTextElementAttributes, XSLTransformElementAttributes, XSLValueOfElementAttributes, XSLVariableElementAttributes, XSLWhenElementAttributes, XSLWithParamElementAttributes } from "@zfeed/xsl";

function xsl<T extends Record<string, any>>(name: string) {
    return (props: T, { slots }: Omit<SetupContext, "expose">) => h(`xsl:${name}`, props, slots.default?.());
}

export const XslApplyImport = xsl<XSLApplyImportElementAttributes>("apply-import");
export const XslApplyTemplates = xsl<XSLApplyTemplatesElementAttributes>("apply-templates");
export const XslAttributeSet = xsl<XSLAttributeSetElementAttributes>("attribute-set");
export const XslAttribute = xsl<XSLAttributeElementAttributes>("attribute");
export const XslCallTemplate = xsl<XSLCallTemplateElementAttributes>("call-template");
export const XslChoose = xsl<XSLChooseElementAttributes>("choose");
export const XslComment = xsl<XSLCommentElementAttributes>("comment");
export const XslCopyOf = xsl<XSLCopyOfElementAttributes>("copy-of");
export const XslCopy = xsl<XSLCopyElementAttributes>("copy");
export const XslDecimalFormat = xsl<XSLDecimalFormatElementAttributes>("decimal-format");
export const XslElement = xsl<XSLElementElementAttributes>("element");
export const XslFallback = xsl<XSLFallbackElementAttributes>("fallback");
export const XslForEach = xsl<XSLForEachElementAttributes>("for-each");
export const XslIf = xsl<XSLIfElementAttributes>("if");
export const XslImport = xsl<XSLImportElementAttributes>("import");
export const XslInclude = xsl<XSLIncludeElementAttributes>("include");
export const XslKey = xsl<XSLKeyElementAttributes>("key");
export const XslMessage = xsl<XSLMessageElementAttributes>("message");
export const XslNamespaceAlias = xsl<XSLNamespaceAliasElementAttributes>("namespace-alias");
export const XslNumber = xsl<XSLNumberElementAttributes>("number");
export const XslOtherwise = xsl<XSLOtherwiseElementAttributes>("otherwise");
export const XslOutput = xsl<XSLOutputElementAttributes>("output");
export const XslParam = xsl<XSLParamElementAttributes>("param");
export const XslPreserveSpace = xsl<XSLPreserveSpaceElementAttributes>("preserve-space");
export const XslProcessingInstruction = xsl<XSLProcessingInstructionElementAttributes>("processing-instruction");
export const XslSort = xsl<XSLSortElementAttributes>("sort");
export const XslStripSpace = xsl<XSLStripSpaceElementAttributes>("strip-space");
export const XslStylesheet = xsl<XSLStylesheetElementAttributes>("stylesheet");
export const XslTemplate = xsl<XSLTemplateElementAttributes>("template");
export const XslText = xsl<XSLTextElementAttributes>("text");
export const XslTransform = xsl<XSLTransformElementAttributes>("transform");
export const XslValueOf = xsl<XSLValueOfElementAttributes>("value-of");
export const XslVariable = xsl<XSLVariableElementAttributes>("variable");
export const XslWhen = xsl<XSLWhenElementAttributes>("when");
export const XslWithParam = xsl<XSLWithParamElementAttributes>("with-param");
