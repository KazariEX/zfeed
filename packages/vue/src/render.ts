import { renderToString } from "@vue/server-renderer";
import { type Component, createSSRApp } from "vue";

/**
 * Copy from `vue-component-type-helpers`
 *
 * @link https://github.com/vuejs/language-tools/blob/v2.2.10/packages/component-type-helpers/index.ts#L6
 */
type ComponentProps<T> = T extends new (...args: any) => {
    $props: infer P;
} ? NonNullable<P> : T extends (props: infer P, ...args: any) => any ? P : {};

export async function render<T extends Component>(component: T, props?: ComponentProps<T>) {
    const app = createSSRApp(component, props);
    const text = await renderToString(app);
    return `<?xml version="1.0" encoding="UTF-8"?>${text}`;
}
