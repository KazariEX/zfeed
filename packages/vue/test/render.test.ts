import { html as format } from "js-beautify";
import { describe, expect, it } from "vitest";
import { render } from "../src";
import Comp from "./fixtures/index.vue";

describe("render", () => {
    it("snapshot", async () => {
        const text = await render(Comp);
        expect(format(text)).toMatchSnapshot();
    });
});
