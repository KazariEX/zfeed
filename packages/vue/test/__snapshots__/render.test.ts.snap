// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render > snapshot 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform" xmlns:atom="http://www.w3.org/2005/Atom">
    <xsl:output method="html"></xsl:output>
    <xsl:template match="/">
        <html lang="{atom:feed/atom:language}">

        <head>
            <meta charset="utf-8">
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <title><xsl:value-of select="atom:feed/atom:title/"></xsl:value-of></title>
            <link rel="stylesheet" href="/assets/atom.css">
            <link rel="icon" href="{atom:feed/atom:icon}">
        </head>

        <body>
            <main><xsl:apply-templates select="atom:feed/atom:entry"></xsl:apply-templates></main>
        </body>

        </html>
    </xsl:template>
    <xsl:template match="atom:entry"><a href="{atom:link/@href}">
            <xsl:variable name="img-src" select="substring-before(substring-after(substring-after(atom:content, &#39;&lt;img&#39;), &#39;src=&quot;&#39;), &#39;&quot;&#39;)"></xsl:variable>
            <xsl:if test="$img-src"><img class="entry-image" src="{$img-src}" alt="{atom:title}" loading="lazy"></xsl:if>
            <article>
                <h2 class="entry-title"><xsl:value-of select="atom:title"></xsl:value-of></h2>
                <xsl:if test="atom:summary">
                    <div class="entry-summary"><xsl:value-of select="atom:summary"></xsl:value-of></div>
                </xsl:if>
                <div class="entry-meta">
                    <xsl:variable name="published-date" select="substring(atom:published, 1, 10)"></xsl:variable> 发布于 <xsl:value-of select="$published-date"></xsl:value-of>
                    <xsl:if test="atom:updated and atom:updated != atom:published">
                        <xsl:text> • 更新于 </xsl:text><xsl:value-of select="substring(atom:updated, 1, 10)"></xsl:value-of>
                    </xsl:if>
                    <xsl:if test="atom:category">
                        <xsl:text> • </xsl:text><xsl:for-each select="atom:category"><xsl:value-of select="@term"></xsl:value-of></xsl:for-each>
                    </xsl:if>
                </div>
            </article>
        </a></xsl:template>
</xsl:stylesheet>"
`;
